### PR TITLE
Detect Wine and disable unsupported IAudioClient3 interface.

### LINKED ISSUE #67327: v4 Beta 2 Win64 - Windows Desktop exported EXE Does Not Run On Linux WINE?
### Godot version

v4.0.beta2.official [f8745f2f7]

### System information

Windows 11 Pro 64Bit/Linux Mint 21 Cinnamon 64Bit

### Issue description

Hi,

Version: v4.0.beta2.official [f8745f2f7] does not build
a Windows Desktop EXE that runs on Linux WINE version 7.x ?

Can this be fixed in Beta 3 ?

Jesse

### Steps to reproduce

Run Godot v4.0.beta2.official [f8745f2f7]
On Windows 11 Pro 64Bit
Then export game to Windows Desktop EXE executable.
Copy game folder to a Linux and run in WINE 6.x+.
WINE crashes:

________________________________________________________________
```
Unhandled exception: page fault on read access to 0x0000000000000060 in 64-bit code (0x000000014262bbb7).
Register dump:
 rip:000000014262bbb7 rsp:000000000081f260 rbp:000000000081f8f0 eflags:00010206 (  R- --  I   - -P- )
 rax:0000000000000000 rbx:0000000009da85e0 rcx:0000000000000000 rdx:0000000009d40000
 rsi:0000000142599690 rdi:0000000009da85f8  r8:0000000009da89f0  r9:0000000000000000 r10:0000000009f60000
 r11:0000000009da8b78 r12:000000000081f320 r13:000000000081f310 r14:0000000009da85e8 r15:0000000009da89f0
Stack dump:
0x000000000081f260:  000000000081f348 000000017002a77b
0x000000000081f270:  0000000000000002 00000001429862ef
0x000000000081f280:  000000000081f338 000000014000cf99
0x000000000081f290:  0000000000000001 00000001c8dd5c5e
0x000000000081f2a0:  0000000000000002 0000000000000002
0x000000000081f2b0:  0000000000000002 0000000000000002
0x000000000081f2c0:  000000000081f328 00000001429b2f44
0x000000000081f2d0:  000000000081f8f0 000000000081f9b0
0x000000000081f2e0:  0000000143709fa0 000000000081f990
0x000000000081f2f0:  000000000081f970 00000001c8dd5c84
0x000000000081f300:  0000000009da8470 0000000000000000
0x000000000081f310:  0000000000000000 0000000000000005
Backtrace:
=>0 0x000000014262bbb7 EntryPoint+0x262a6f7() in tetristory (0x000000000081f8f0)
  1 0x0000000142df67b9 EntryPoint+0x2df52f9() in tetristory (0x000000000081f8f0)
  2 0x000000014006c9b6 EntryPoint+0x6b4f6() in tetristory (0x000000000085b320)
  3 0x00000001430ef0d5 EntryPoint+0x30edc15() in tetristory (0x0000000000000001)
  4 0x00000001400013c1 in tetristory (+0x13c1) (0x0000000000000001)
  5 0x00000001400014d6 EntryPoint+0x16() in tetristory (0x0000000000000000)
  6 0x000000007b62c599 ActivateActCtx+0x20a19() in kernel32 (0x0000000000000000)
  7 0x0000000170059d53 A_SHAFinal+0x37e63() in ntdll (0x0000000000000000)
0x000000014262bbb7 tetristory+0x262bbb7: movq	0x0000000000000060(%r9),%rcx
Modules:
Module	Address					Debug info	Name (36 modules)
PE	000000007a850000-000000007a854000	Deferred        opengl32
PE	000000007b000000-000000007b0d5000	Deferred        kernelbase
PE	000000007b600000-000000007b812000	Export          kernel32
PE	0000000140000000-0000000143e03000	Export          tetristory
PE	0000000170000000-000000017009a000	Export          ntdll
PE	00000001c69e0000-00000001c72fc000	Deferred        shell32
PE	00000001c8b40000-00000001c8b60000	Deferred        msacm32
PE	00000001c8db0000-00000001c8e46000	Deferred        msvcrt
PE	00000001d7cb0000-00000001d7cc1000	Deferred        wsock32
PE	00000001ec2b0000-00000001ec2d5000	Deferred        ws2_32
PE	00000001f51e0000-00000001f51f0000	Deferred        hid
PE	000000021a7e0000-000000021a855000	Deferred        setupapi
PE	0000000231ae0000-0000000231b62000	Deferred        rpcrt4
PE	000000023d820000-000000023da6a000	Deferred        user32
PE	0000000240030000-000000024005e000	Deferred        iphlpapi
PE	000000025d740000-000000025d74e000	Deferred        dwmapi
PE	000000026b4c0000-000000026b53b000	Deferred        gdi32
PE	000000028dfa0000-000000028dfac000	Deferred        nsi
PE	000000029cfc0000-000000029cfd6000	Deferred        dnsapi
PE	00000002bb750000-00000002bb890000	Deferred        comctl32
PE	00000002d4d40000-00000002d4d56000	Deferred        bcrypt
PE	00000002e3540000-00000002e3591000	Deferred        shlwapi
PE	00000002e8f10000-00000002e902b000	Deferred        ole32
PE	00000002f1fa0000-00000002f1fad000	Deferred        version
PE	00000002f2930000-00000002f293c000	Deferred        avrt
PE	0000000308050000-000000030809b000	Deferred        dinput8
PE	000000030a950000-000000030a9c1000	Deferred        dwrite
PE	00000003126f0000-0000000312709000	Deferred        shcore
PE	0000000327020000-0000000327072000	Deferred        combase
PE	000000032a700000-000000032a729000	Deferred        sechost
PE	0000000330260000-000000033029f000	Deferred        advapi32
PE	0000000375610000-0000000375648000	Deferred        win32u
PE	00000003af670000-00000003af72f000	Deferred        ucrtbase
PE	00000003afd00000-00000003afd1a000	Deferred        imm32
PE	00000003b8f00000-00000003b8fc1000	Deferred        winmm
PE	00007f11f4080000-00007f11f4084000	Deferred        winex11
Threads:
process  tid      prio (all id:s are in hex)
00000038 services.exe
	0000003c    0
	00000040    0
	0000004c    0
	00000050    0
	00000074    0
	00000098    0
	000000b0    0
	000000d4    0
	000000d8    0
00000044 winedevice.exe
	00000048    0
	00000054    0
	00000058    0
	0000005c    0
	00000060    0
	000000bc    0
00000064 winedevice.exe
	00000068    0
	00000078    0
	0000007c    0
	00000080    0
	00000084    0
	00000088    0
	0000008c    0
0000006c explorer.exe
	00000070    0
	000000c0    0
	000000c4    0
00000090 plugplay.exe
	00000094    0
	0000009c    0
	000000a0    0
	000000a4    0
000000a8 svchost.exe
	000000ac    0
	000000b4    0
	000000b8    0
000000cc rpcss.exe
	000000d0    0
	000000dc    0
	000000e0    0
	000000e4    0
	000000e8    0
	000000ec    0
	000000f0    0
000000fc (D) Z:\home\jlp\Desktop\TS3-G4-Win64\TetriStory.exe
	00000100    0 <==
	00000104    0
	00000108    0
	0000010c    0
	00000110    0
	00000114    0
	00000118    0
	0000011c    0
	00000120    0
	00000124    0
	00000128    0
	0000012c    0
	00000130    0
	00000134    0
	00000138    0
	0000013c    0
	00000140    0
	00000144    0
	00000150    0
System information:
    Wine build: wine-7.0
    Platform: x86_64
    Version: Windows 7
    Host system: Linux
    Host version: 5.15.0-50-generic
```

________________________________________________________________

### Minimal reproduction project

https://github.com/BetaMaxHero/GDScript_Godot_4_T-Story

### PR DESCRIPTION
**Pull request by bruvzg**
Fixes #67327

### CODE REVIEW

File: drivers/wasapi/audio_driver_wasapi.cpp [hunk 1]
```diff
@@ -201,6 +206,26 @@ class CMMNotificationClient : public IMMNotificationClient {
...
+typedef const char *(CDECL *PWineGetVersionPtr)(void);
+
+bool AudioDriverWASAPI::is_running_on_wine() {
+	HMODULE nt_lib = LoadLibraryW(L"ntdll.dll");
+	if (!nt_lib) {
+		return false;
+	}
+
+	PWineGetVersionPtr wine_get_version = (PWineGetVersionPtr)GetProcAddress(nt_lib, "wine_get_version");
+	if (wine_get_version) {
```
Comment (akien-mga) [inline drivers/wasapi/audio_driver_wasapi.cpp:218]: Could factor out freeing the lib, and then maybe even skipping the if/else with `return (bool)wine_get_version;`.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by unfa** [comment]
If it's possible to detect when running in Wine, can I make my game tell the user to download a native version for optimal experience?

**Comment by Calinou** [comment]
> If it's possible to detect when running in Wine, can I make my game tell the user to download a native version for optimal experience?

The method is not exposed to scripting, but I'd think twice before doing this. The user may have good reasons to be running the app/game in WINE (e.g. using [ReShade](https://reshade.me/), which can't be used on native Linux apps but can be used on Windows apps via WINE).

**Comment by unfa** [comment]
I see. Well, I wouldn't enforce it, but for example show a popup once in case someone did that by mistake.
Also - isn't vkBasalt allowing to do something similar for native Linux software?

**Pull request merged**